src/app/qgssubstitutionlistwidget.cpp

```diff
@@ -0,0 +1,230 @@
+{
+  QList< QgsStringReplacement > result;
+  for ( int i = 0; i < mTableSubstitutions->rowCount(); ++i )
+  {
+    if ( !mTableSubstitutions->item( i, 0 ) )
+      continue;
+
+    if ( mTableSubstitutions->item( i, 0 )->text().isEmpty() )
+      continue;
+
+    QCheckBox* chkCaseSensitive = qobject_cast<QCheckBox*>( mTableSubstitutions->cellWidget( i, 2 ) );
+    QCheckBox* chkWholeWord = qobject_cast<QCheckBox*>( mTableSubstitutions->cellWidget( i, 3 ) );
+
+    QgsStringReplacement replacement( mTableSubstitutions->item( i, 0 )->text(),
+                                      mTableSubstitutions->item( i, 1 )->text(),
+                                      chkCaseSensitive->isChecked(),
+                                      chkWholeWord->isChecked() );
+    result << replacement;
+  }
+  return QgsStringReplacementCollection( result );
+}
+
+void QgsSubstitutionListWidget::on_mButtonAdd_clicked()
+{
+  addSubstitution( QgsStringReplacement( QString(), QString(), false, true ) );
+  mTableSubstitutions->setFocus();
+  mTableSubstitutions->setCurrentCell( mTableSubstitutions->rowCount() - 1, 0 );
+}
+
+void QgsSubstitutionListWidget::on_mButtonRemove_clicked()
+{
+  int currentRow = mTableSubstitutions->currentRow();
+  mTableSubstitutions->removeRow( currentRow );
+  tableChanged();
+}
+
+void QgsSubstitutionListWidget::tableChanged()
+{
+  emit substitutionsChanged( substitutions() );
+}
+
+void QgsSubstitutionListWidget::on_mButtonExport_clicked()
+{
+  QString fileName = QFileDialog::getSaveFileName( this, tr( "Save substitutions" ), QDir::homePath(),
+                     tr( "XML files (*.xml *.XML)" ) );
+  if ( fileName.isEmpty() )
+  {
+    return;
+  }
+
+  // ensure the user never ommited the extension from the file name
+  if ( !fileName.endsWith( ".xml", Qt::CaseInsensitive ) )
+  {
+    fileName += ".xml";
+  }
+
+  QDomDocument doc;
+  QDomElement root = doc.createElement( "substitutions" );
+  root.setAttribute( "version", "1.0" );
+  QgsStringReplacementCollection collection = substitutions();
+  collection.writeXml( root, doc );
+  doc.appendChild( root );
+
+  QFile file( fileName );
+  if ( !file.open( QIODevice::WriteOnly | QIODevice::Text ) )
+  {
+    QMessageBox::warning( nullptr, tr( "Export substitutions" ),
+                          tr( "Cannot write file %1:\n%2." ).arg( fileName, file.errorString() ),
+                          QMessageBox::Ok,
+                          QMessageBox::Ok );
+    return;
+  }
+
+  QTextStream out( &file );
+  doc.save( out, 4 );
+}
+
+void QgsSubstitutionListWidget::on_mButtonImport_clicked()
+{
+  QString fileName = QFileDialog::getOpenFileName( this, tr( "Load substitutions" ), QDir::homePath(),
+                     tr( "XML files (*.xml *.XML)" ) );
+  if ( fileName.isEmpty() )
+  {
+    return;
+  }
+
+  QFile file( fileName );
+  if ( !file.open( QIODevice::ReadOnly | QIODevice::Text ) )
+  {
+    QMessageBox::warning( nullptr, tr( "Import substitutions" ),
```

Comment:
I usually would, but in this case the errors might be encountered within the layer properties dialog. I'm not sure showing the errors in the canvas is very friendly in that case.